extensions/notebook/src/book/remoteBookController.ts

```diff
@@ -0,0 +1,146 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as request from 'request';
+import * as loc from '../common/localizedConstants';
+import * as utils from '../common/utils';
+import * as vscode from 'vscode';
+import { RemoteBookDialogModel } from '../dialog/remoteBookDialog';
+import { GitHubRemoteBook } from '../book/githubRemoteBook';
+import { SharedRemoteBook } from '../book/sharedRemoteBook';
+
+const assetNameRE = /([a-zA-Z0-9]+)(?:-|_)([a-zA-Z0-9.]+)(?:-|_)([a-zA-Z0-9]+).(zip|tar.gz|tgz)/;
+
+export class RemoteBookController {
+	constructor(public model: RemoteBookDialogModel, public outputChannel: vscode.OutputChannel) {
+	}
+
+	public async setRemoteBook(url: URL, remoteLocation: string, asset?: IAsset): Promise<void> {
+		if (remoteLocation === 'GitHub') {
+			this.model.remoteBook = new GitHubRemoteBook(url, this.outputChannel, asset);
+		} else {
+			this.model.remoteBook = new SharedRemoteBook(url, this.outputChannel);
+		}
+		return await this.model.remoteBook.createLocalCopy();
+	}
+
+	public async getReleases(url?: URL): Promise<IRelease[]> {
+		if (url) {
+			this.model.releases = [];
+			let options = {
+				headers: {
+					'User-Agent': 'request'
+				}
+			};
+			return new Promise<IRelease[]>((resolve, reject) => {
+				request.get(url.href, options, (error, response, body) => {
+					if (error) {
+						return reject(error);
+					}
+
+					if (response.statusCode !== 200) {
+						return reject(new Error(loc.httpRequestError(response.statusCode, response.statusMessage)));
+					}
+
+					let releases = JSON.parse(body);
+					let bookReleases: IRelease[] = [];
+					if (releases?.length > 0) {
+						let keys = Object.keys(releases);
+						keys.forEach(key => {
+							try {
+								bookReleases.push({ name: releases[key].name, assetsUrl: new URL(releases[key].assets_url) });
+							}
+							catch (error) {
+								return reject(error);
+							}
+						});
+					}
+					if (bookReleases.length > 0) {
+						this.model.releases = bookReleases;
+						resolve(bookReleases);
+					} else {
+						return reject(new Error(loc.msgReleaseNotFound));
+					}
+				});
+			});
+		} else {
+			return new Promise<IRelease[]>((resolve, reject) => {
```

Comment:
Just `return this.model.releases` - async functions will wrap it in a promise for you.